edgelet/edgelet-http/src/certificate_manager.rs

```diff
@@ -0,0 +1,243 @@
+#![deny(rust_2018_idioms, warnings)]
+#![deny(clippy::all, clippy::pedantic)]
+
+use std::sync::{Arc, RwLock};
+
+#[cfg(unix)]
+use openssl::pkcs12::Pkcs12;
+#[cfg(unix)]
+use openssl::pkey::PKey;
+#[cfg(unix)]
+use openssl::stack::Stack;
+#[cfg(unix)]
+use openssl::x509::X509;
+
+use edgelet_core::crypto::{Certificate, CreateCertificate, KeyBytes, PrivateKey, Signature};
+use edgelet_core::CertificateProperties;
+use failure::ResultExt;
+
+pub use crate::error::{Error, ErrorKind};
+
+#[derive(Clone)]
+pub struct CertificateManager<C: CreateCertificate + Clone> {
+    certificate: Arc<RwLock<Option<(String, String)>>>,
+    crypto: C,
+    props: CertificateProperties,
+}
+
+impl<C: CreateCertificate + Clone> CertificateManager<C> {
+    pub fn new(crypto: C, props: CertificateProperties) -> Self {
+        Self {
+            certificate: Arc::new(RwLock::new(None)),
+            crypto,
+            props,
+        }
+    }
+
+    // Convenience function since native-tls does not yet support PEM
+    // and since everything else uses PEM certificates, we want to keep
+    // the actual storage of the certificate in the PEM format.
+    #[cfg(unix)]
+    pub fn get_pkcs12_certificate(&self) -> Result<Vec<u8>, Error> {
+        let stored_cert_bundle = self
+            .get_certificate()
+            .with_context(|_| ErrorKind::CertificateCreationError)?;
```

Comment:
Since `get_certificate` already returns `Error` this `with_context` may not be necessary.

Reply:
See above.